### PR TITLE
Microchip: MEC172x: Add headers 1

### PR DESCRIPTION
**Pull request by scottwcpg**
Add first three MEC172x in-tree register headers. This PR part of series and depends on PR 36554 Microchip MEC172x initial submission.
These headers are formatted per Zephyr coding standards and do not contain device addresses only register offsets.  MEC172x device tree will provide all peripheral address.

---
This change is [Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36654)

### CODE REVIEW

**Comment by scottwcpg** [comment]
Sigh. I created a sub-branch, added some of the new headers, commit, and the push.  Git added the sub-branch to this PR.  I expected a new PR.  Could someone please enlighten me on what I did wrong?

**Comment by scottwcpg** [comment]
Hi @sjg20 , @VenkatKotakonda , @albertofloyd 
Forgive my lack of ability with git and github.  I tried creating a sub-branch to add a few headers but I did not do something correct and git/github did not create a new PR.  Instead it added the header commit to this PR and cleared the code review approvals. Any pointers to documentation that explains step by step how to get separate PR's from sub branches?

**Comment by scottwcpg** [comment]
Hi @sjg20 , @albertofloyd  No code owner was set for soc/arm/microchip_mec and it is defaulting to the default owners for the whole soc/arm tree.  Can we request a code owner update for soc/arm/microchip_mec?  Most of the other EC's have owners.

**Review by galak — COMMENTED** [review]
Are the headers in mec172x/reg/* something that there is some other "upstream" source you'd be syncing with.

I ask as I still think it makes more sense that peripheral structs and register defines are kept with the drivers that utilize them and only things you need for code that uses it under soc/arm/microchip_mec/mec172x/ would be in soc/arm/microchip_mec/mec172x/reg*

**Comment by scottwcpg** [comment]
Hi @galak 
I can agree with you if the header files containing these structures can be included from SoC, other drivers, or application level.  Are headers in drivers/<periph>/ expected to be private or can they be included by other code?  Many HW block are interdependent.  
Many peripherals do not yet have drivers or do not need drivers.  Is is OK to put headers without drivers in the drivers/<periph> hierarchy or add new drivers/<periph> folders?

**Comment by scottwcpg** [comment]
Hi @galak @sjg20 , @albertofloyd @VenkatKotakonda 
In response to the request to put register structures and their bit definitions at the driver level, I observed STM and other vendors have headers in zephyr/include/drivers/<peripheral>.
The MEC172x in these PR currently needs: MEC ECIA, PCR, ECS, VBATR, and GPIO structures and defines.
I propose to put headers in 
zephyr/include/drivers/clock_control/ for PCR and VBATR these HW block have clock related stuff
zephyr/include/drivers/interrupt_controller/ for ECIA it is an interrupt related block.
zephyr/include/drivers/gpio/ for MEC GPIO stuff
Future headers for register structures and defines will go into the other relevant folders in zephyr/include/drivers/
Is this acceptable?

**Review by sjg20 — APPROVED** [review]

**Review by keith-zephyr — COMMENTED** [review]
I think the proposed organization of Scott's last comment is good. (PCR, VBATR into clock_control/, ECIA into interrupt_controller/, GPIO in gpio/)

> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/36654#-:-MddDhP_ECceBnXffKar:b-yl8yx7)* status: 0 of 6 files reviewed, all discussions resolved (waiting on @albertofloyd, @ioannisg, @MaureenHelm, and @VenkatKotakonda)

File: soc/arm/microchip_mec/mec172x/reg/mec_vbat.h [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _MEC_VBAT_H
+#define _MEC_VBAT_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+/* VBAT Registers Registers */
```
Comment (VenkatKotakonda) [inline soc/arm/microchip_mec/mec172x/reg/mec_vbat.h:13]: Do we have BGPO register definitions?
i do see mec172x has BGPO

File: soc/arm/microchip_mec/mec172x/reg/mec_gpio.h [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+	MCHP_GPIO_DIR_OUT,
+};
+
+enum mchp_gpio_parout_en {
+	MCHP_GPIO_PAROUT_DIS = 0,
+	MCHP_GPIO_PAROUT_EN,
+};
+
+enum mchp_gpio_pol {
+	MCHP_GPIO_POL_NORM = 0,
```
Comment (VenkatKotakonda) [inline soc/arm/microchip_mec/mec172x/reg/mec_gpio.h:353]: some unsigned numbers has "u" postfix and some doesnt.
Please fix it.

**Comment by scottwcpg** [comment]
I'm at the breaking point.  I do not know what to do with this code to make everyone happy.  I have worked through memorial day, my anniversary in June, and now the 4th of July holiday.  Can't do this anymore.

**Comment by albertofloyd** [comment]
> get separate PR's from sub branches?

https://guides.github.com/introduction/flow/

**Comment by albertofloyd** [comment]
> Hi @sjg20 , @albertofloyd No code owner was set for soc/arm/microchip_mec and it is defaulting to the default owners for the whole soc/arm tree. Can we request a code owner update for soc/arm/microchip_mec? Most of the other EC's have owners.

@scottwcpg You can create a PR modifying CODEOWNERS like this https://github.com/zephyrproject-rtos/zephyr/pull/36564

**Comment by albertofloyd** [comment]
> what to do with this code to make everyone happy. I have worked through memorial day, my anniversary in June, and now the 4th of July holiday. Can't do this anymore.

@scottwcpg I understand the pain, the original guideline worked well just to keep all these headers as hal module but with different opinions assuming guidelines are not to be followed and deviations been considered ok without consulting the rest of community, we ended up here.  I still do not follow how the CMSIS header dependency will be sorted once these headers are here.

You can try adding the header as you are suggesting, unsure if the owners on these folders would allow it.

**Comment by scottwcpg** [comment]
I have fought the Zephyr build system and lose.  I tried to make V2 drivers so the MEC152x code would not be touched.  Failed. I didn't know the build system looks at yaml files names and compatibility names and you must put versioning on everything and it still failed with hundreds of device tree errors.  Does this mean you must create new drivers with completely different YAML compable naming?  I tried to describe the interrupt aggregator using child nodes because it looks clean and the DTSI is clear and understandable.  Three solid days of fighting device tree. Adding info to each DT node means changing existing YAML or creating new YAML just to have a node for each HW block even if you don't need a driver.  On top of that, where do I put headers that have subtle differences between chips?  PCR and VBAT HW blocks have significant differences from MEC152x.  Multiple private header in drivers/clock_control or somewhere else.  SoC specific driver API appear to be in include/drivers/<block> but most of these headers are function prototypes no SoC specific definitions.  ADC has extra registers, QMSPI has significant differences.  No guidance by looking at other EC's they all use HAL's.  Finally, having touched EVERY driver I expect code review rejects.  I can't even test all the changes.  All these changes must be together to test it.  How to stage all these changes?

**Comment by sjg20** [comment]
Hi Scott,

We can probably help provide some technical guidance here, although I think
I'll need to leave the work/life balance issues to you and your company to
figure out :-)

My understanding is that Keith is happy with your proposal and I've already
approved the PR.

As a comment to others, I'd like to see us get some runs on the board and
iterate towards full support, rather than trying to get everything in at
once, or everything perfect up front.

With this PR in, you should be able to pick a subsystem to submit, such as
I2C.

Regards,
SImon

**Pull request closed without merging**